### PR TITLE
#2663 Quarkus Extension for Kafka Streams

### PR DESCRIPTION
**Pull request by gunnarmorling**
~~Planning some more testing under native, but the basic functionality is there.~~
Feedback welcome!
Ready for (re-)review from my side.

### CODE REVIEW

File: extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsSubstitutions.java
Comment (gunnarmorling) [inline extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsSubstitutions.java]: @Sanne, this is the field I'd like to reset. I tried to substitute the class as you suggested, but it seems I can't as it is an enum type:

    com.oracle.svm.core.util.UserError$UserException: Annotated class must inherit directly from Object: class io.quarkus.kafka.streams.runtime.graal.Target_org_rocksdb_RocksDB$Target_org_rocksdb_RocksDB_LibraryState

So subsituting the entire method that makes use of the field is the only way I've found so far. If there's a better/easier one, that'd be great.

**Comment by kenfinnigan** [comment]
I'm quite the noob when it comes to Kafka.

How does this differ from Reactive Messaging with Kafka?

**Comment by gunnarmorling** [comment]
Kafka Streams is used for implementing streaming queries based on top of Kafka, like mapping and filtering topics, grouping/aggregating values, joining topics, windowing etc. It also lets you run interactive queries (querying current state instead of streaming topics), all that in a distributed and scalable fashion.

It's a popular API of the Kafka eco-system (and part of the Kafka project) and having support for it in Quarkus will complete the Kafka story of it. mP Reactive Operators spec can do _some_ of these things, but I don't think it goes as far.

**Comment by gunnarmorling** [comment]
@cescoffier, @gsmet, any feedback? Btw. how can I actually "request a review"? Seems I'd have to be part of the org or something? Thanks!

Btw. I did some more testing locally; interactive queries work, too. I'll push a test case for this. With that, I'd consider it good enough for a first release. Happy to learn about your thoughts.

File: extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsSubstitutions.java
Comment (Sanne) [reply extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsSubstitutions.java]: hey @gunnarmorling , Substitutions are indeed a bit limited. They are also a bitch to maintain in the long term, so don't look at them via your perfectionist hat, but war the pragmatic one; personally I've used them to speed-up things as otherwise one ends up in the neverending loop of waiting for 3rd parties to accept PRs and wait for all such releases to align.

So replace all the method and don't feel too bad about it :)

Either way, the proper solution is to craft a PR for the upstream project, so to not live with such stuff for too long. Sometimes that's not entirely possible, but you could e.g. refactor the upstream project to minimize the scope of substitutions.

e.g. PostgreSQL JDBC no longer needs substitutions , the ones we have for Hibernate ORM are much smaller/simpler now than what they used to be (and should reduce further with further ORM work), the ones for Narayana are about to become no longer necessary either, etc...

Comment (Sanne) [reply extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsSubstitutions.java]: P.S. to answer your question: there might be a better way, but I don't know it :) Either way, suggesting to not bother.

Comment (gunnarmorling) [reply extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsSubstitutions.java]: Thanks, @Sanne! So one alternative would be to delay initialization of all the classes calling this method in the static initializer to runtime. This would avoid the substitution here mostly. I was originally shying away from it, but it might be better indeed. What's the "Quarkus way" of doing this?

Comment (Sanne) [reply extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsSubstitutions.java]: it depends on your needs. Generally we try to "run at build" if that would imply either better dead-code-elimination possibilities (e.g. the constant being resolved having some effect like disabling a feature) or if it would otherwise speed-up the runtime bootstrap (by avoiding doing some work at runtime).

If you have neither such benefits, might be better to do such things at runtime. Especially if you have this code need to re-run at runtime, then there's no need to bother running it during the build.

Comment (gunnarmorling) [reply extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsSubstitutions.java]: Yes, so how do I configure a class to be initialized at runtime with Quarkus then?

File: integration-tests/kafka/src/test/java/io/quarkus/it/main/KafkaStreamsTest.java
Comment (gsmet) [inline integration-tests/kafka/src/test/java/io/quarkus/it/main/KafkaStreamsTest.java]: @cescoffier I let you check if this is consistent with how Kafka things are configured in the other extensions.

File: extensions/kafka-streams/deployment/pom.xml
Comment (gsmet) [inline extensions/kafka-streams/deployment/pom.xml]: I'm surprised you need `undertow` and `jsonp`? Are they required by `kafka-streams`?

You don't have the runtime dependencies for them so I suppose they are not and it's just something you missed?

As for `arc`, you don't have the runtime dependency either. So I would either remove it from deployment here or add it as runtime too.

**Review by gsmet — DISMISSED** [review]
Code looks good. I added a comment about some dependencies that look incorrect and the configuration where I think we need @cescoffier to step in.

BTW, we should probably either extend the existing Kafka guide (it's probably already complex enough) or write another one if we think it's worth advertising this feature.

File: extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsSubstitutions.java
Comment (gunnarmorling) [reply extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/graal/KafkaStreamsSubstitutions.java]: Never mind, I've found what I need in the Substrate build items.

File: integration-tests/kafka/src/test/java/io/quarkus/it/main/KafkaStreamsTest.java
Comment (gunnarmorling) [reply integration-tests/kafka/src/test/java/io/quarkus/it/main/KafkaStreamsTest.java]: FWIW, it's copied 1:1 from the existing Kafka tests.

File: extensions/kafka-streams/deployment/pom.xml
Comment (gunnarmorling) [reply extensions/kafka-streams/deployment/pom.xml]: Good catch, this was c&p most likely. Removed it and updated the PR.

File: extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java
Comment (gunnarmorling) [inline extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java]: So this does the trick for avoiding all that substitued method I had before. Much happier with it now :)

**Comment by gunnarmorling** [comment]
Ok, so provided CI passes, I'm happy with this change now. All, please let me know whether you see any issues. Thanks!

File: extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java
Comment (Sanne) [reply extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java]: nice! thanks for digging that up :)

File: extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java
Comment (gsmet) [inline extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java]: Do you still need that now that you delayed the initialization at runtime?

**Comment by gsmet** [comment]
I added one additional comment. Otherwise, it looks good code wise.

@cescoffier could you check it's in line with the general strategy you have with Apache Kafka support.

Also a guide/quickstart explaining why it is useful would be nice (let's wait for @cescoffier approval first).

**Comment by cescoffier** [comment]
First to answer to @gunnarmorling most of the features are available in MP Reactive Messaging when you use Stream-based functions (generally using RX Java 2, or Reactor) (windowing, groups, join...). The main differences are:
* it's not Kafka specific
* it does not back things into Kafka (but just in memory, so it's good and bad at the same time)

Now, I agree that Kafka streams are popular in the Kafka world, and we should have it. For Quarkus the main strategy stays MP Reactive Messaging. This extension should be used if you have a Kafka Streams pipeline and want to migrate to Quarkus.

File: integration-tests/kafka/src/main/java/io/quarkus/it/kafka/KafkaStreamsPipeline.java
Comment (cescoffier) [inline integration-tests/kafka/src/main/java/io/quarkus/it/kafka/KafkaStreamsPipeline.java]: can you use awaitiliby?

File: integration-tests/kafka/src/test/java/io/quarkus/it/main/KafkaStreamsTest.java
Comment (cescoffier) [reply integration-tests/kafka/src/test/java/io/quarkus/it/main/KafkaStreamsTest.java]: in raw kafka it's what you do.... yes...

File: integration-tests/kafka/src/test/java/io/quarkus/it/main/KafkaStreamsTest.java
Comment (cescoffier) [inline integration-tests/kafka/src/test/java/io/quarkus/it/main/KafkaStreamsTest.java]: order may differ, JSON does not enforce ordering.

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: devtools/common/src/main/filtered/extensions.json
Comment (gsmet) [reply devtools/common/src/main/filtered/extensions.json]: It's a label used for Search. So we should have kafka anyway (and we can have kafka-streams too)

Comment (gunnarmorling) [reply devtools/common/src/main/filtered/extensions.json]: Yes, that's what I wanted to ask, too. What's the purpose of this label. If it is for searching suitable extensions, I could add both, "kafka" and "kafka-streams".

File: extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java
Comment (gunnarmorling) [reply extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java]: Yes, it's still needed. The class initializer (which calls this method) won't run as all other classes calling it are initialized at build time.

**Comment by gunnarmorling** [comment]
Thanks for reviewing; I'll update for Awaitility and JSON ordering and force-push then.

File: integration-tests/kafka/src/main/java/io/quarkus/it/kafka/KafkaStreamsPipeline.java
Comment (gunnarmorling) [reply integration-tests/kafka/src/main/java/io/quarkus/it/kafka/KafkaStreamsPipeline.java]: Done.

File: integration-tests/kafka/src/test/java/io/quarkus/it/main/KafkaStreamsTest.java
Comment (gunnarmorling) [reply integration-tests/kafka/src/test/java/io/quarkus/it/main/KafkaStreamsTest.java]: Mh, ok. `JsonObjectBuilderImpl` in the GlassFish impl uses a linked hash map, so order is stable; I've changed it into separate assertions nevertheless.

**Comment by gunnarmorling** [comment]
Reworked as requested, rebased and force-pushed.

**Comment by gunnarmorling** [comment]
Also added one more commit to move the integration tests into the "kafka" package, a left-over from #2691.

File: devtools/common/src/main/filtered/extensions.json
Comment (cescoffier) [reply devtools/common/src/main/filtered/extensions.json]: `-Dextension=kafka` must work and add the MP Kafka support. That's why this one can't be kafka.

**Comment by cescoffier** [comment]
@gunnarmorling I've added one comment around the extension tags. That the last remaining bit.

File: devtools/common/src/main/filtered/extensions.json
Comment (gsmet) [reply devtools/common/src/main/filtered/extensions.json]: @cescoffier well, you can discuss with @emmanuelbernard but that's definitely not the direction he wanted us to go.

Otherwise, we won't be able to add proper labels to any extensions.

Comment (cescoffier) [reply devtools/common/src/main/filtered/extensions.json]: using `kafka` will give the list making the user confused... 
Kafka is one of the things where it should work without confusion.

Comment (gunnarmorling) [reply devtools/common/src/main/filtered/extensions.json]: Hum, ok. Unique identification goes a bit against common interpretation of what a "label" is. Also that `-Dextension=kafka` adds MP Kafka instead of the plain Kafka client could come to the surprise of people. But if that's what you folks want here, then be it.

Comment (gsmet) [reply devtools/common/src/main/filtered/extensions.json]: Well. I clearly disagree with @cescoffier. Labels will be used for searching extensions in the UI and for all the others, we use labels that could create confusion. 

That’s what I fought to only use the artifact name (with some clever shortcut to avoid requiring quarkus- for instance). But @emmanuelbernard was against it. 

As long as we accepted this behavior, we must accept potential conflicts. I don’t like it but that’s how it is. 

That being said, I just saw he thought about the recent use case I reported with arc not being installable: https://github.com/quarkusio/quarkus/issues/2737 .

With what he suggests, we could keep kafka as a label. So I would say let’s add it and let’s refine our algorithm to avoid the issue!

Comment (cescoffier) [reply devtools/common/src/main/filtered/extensions.json]: Then, it's a big breaking change as most of our quickstarts and demos are based on short names (called labels so far). Maybe we need to introduce a short name things too. 

About kafka not using the default kafka client, that's an opinionated choice we made.

Comment (emmanuelbernard) [reply devtools/common/src/main/filtered/extensions.json]: I laid down some thoughts around it https://github.com/quarkusio/quarkus/issues/2737
I don't think most were thinking that labels would be globally unique identifiers. I personally saw it as a way for the search to return all relevant things about $label
Now with my proposal to even remove labels from the `-Dextensions` algorithm, it would mean that MP Reactive Messaging would not show up (unless one does `-Dsearch`).

I do understand Clement's point to try and guide Kafka message consumers towards MP Reactive Messaging. But it kinda make sense for kafka streams to have a kafka label. If we assume that people don't know the difference between kafka streams and kafka client, then I think we are doomed on such a minimalist UI as the starter or the `add-extensions` command.

**Comment by gunnarmorling** [comment]
@emmanuelbernard, @cescoffier, did you have a chance to converge on this one during your stand-up yesterday? How should we move forward? I'd like to wrap up this one :) Thanks!

**Comment by cescoffier** [comment]
@gunnarmorling Sorry, forgot to add a link to https://github.com/quarkusio/quarkus/issues/2737 where we are changing the extension selection algorithm.

As this new algorithm is planned for 0.17.0, I believe we can merge this one with the `kafka` label.

**Comment by cescoffier** [comment]
Just because sometimes basics idea are hard to implement I'm giving a spin to the idea from https://github.com/quarkusio/quarkus/issues/2737 and so how it goes...

**Comment by gunnarmorling** [comment]
> As this new algorithm is planned for 0.17.0, I believe we can merge this one with the kafka label.

That's great news!

I'm working on a quickstart example for this extension now btw.

**Comment by cescoffier** [comment]
Great!

**Comment by cescoffier** [comment]
I've opened a PR for the new selection algorithm (#2785). So we can merge this PR.

**Review by cescoffier — APPROVED** [review]

**Comment by gunnarmorling** [comment]
Yeehaw! Thanks everybody.

**Pull request merged**
